src/bun.js/api/bun/x509.zig

```diff
@@ -0,0 +1,553 @@
+                    }
+                }
+            },
+        }
+    }
+    return true;
+}
+
+inline fn printAltName(out: *BoringSSL.BIO, name: [*]const u8, length: usize, utf8: bool, safe_prefix: ?[*]const u8) void {
+    if (isSafeAltName(name, length, utf8)) {
+        // For backward-compatibility, append "safe" names without any
+        // modifications.
+        if (safe_prefix) |prefix| {
+            _ = BoringSSL.BIO_printf(out, "%s:", prefix);
+        }
+        _ = BoringSSL.BIO_write(out, name, @intCast(c_int, length));
+    } else {
+        // If a name is not "safe", we cannot embed it without special
+        // encoding. This does not usually happen, but we don't want to hide
+        // it from the user either. We use JSON compatible escaping here.
+        _ = BoringSSL.BIO_write(out, "\"", 1);
+        if (safe_prefix) |prefix| {
+            _ = BoringSSL.BIO_printf(out, "%s:", prefix);
+        }
+        for (0..length) |j| {
+            const c = name[j];
+            if (c == '\\') {
+                _ = BoringSSL.BIO_write(out, "\\\\", 2);
+            } else if (c == '"') {
+                _ = BoringSSL.BIO_write(out, "\\\"", 2);
+            } else if ((c >= ' ' and c != ',' and c <= '~') or (utf8 and (c & 0x80) != 0)) {
+                // Note that the above condition explicitly excludes commas, which means
+                // that those are encoded as Unicode escape sequences in the "else"
+                // block. That is not strictly necessary, and Node.js itself would parse
+                // it correctly either way. We only do this to account for third-party
+                // code that might be splitting the string at commas (as Node.js itself
+                // used to do).
+                _ = BoringSSL.BIO_write(out, bun.cast([*]const u8, &c), 1);
+            } else {
+                // Control character or non-ASCII character. We treat everything as
+                // Latin-1, which corresponds to the first 255 Unicode code points.
+                const hex = "0123456789abcdef";
+                const u = [_]u8{ '\\', 'u', '0', '0', hex[(c & 0xf0) >> 4], hex[c & 0x0f] };
+                _ = BoringSSL.BIO_write(out, &u, @sizeOf(@TypeOf(u)));
+            }
+        }
+        _ = BoringSSL.BIO_write(out, "\"", 1);
+    }
+}
+
+inline fn printLatin1AltName(out: *BoringSSL.BIO, name: *BoringSSL.ASN1_IA5STRING, safe_prefix: ?[*]const u8) void {
+    printAltName(out, bun.cast([*]const u8, name.data), @intCast(usize, name.length), false, safe_prefix);
+}
+
+inline fn printUTF8AltName(out: *BoringSSL.BIO, name: *BoringSSL.ASN1_UTF8STRING, safe_prefix: ?[*]const u8) void {
+    printAltName(out, bun.cast([*]const u8, name.data), @intCast(usize, name.length), true, safe_prefix);
+}
+
+pub const kX509NameFlagsRFC2253WithinUtf8JSON = BoringSSL.XN_FLAG_RFC2253 & ~BoringSSL.ASN1_STRFLGS_ESC_MSB & ~BoringSSL.ASN1_STRFLGS_ESC_CTRL;
+
+const NID_ms_upn = 649;
+const NID_id_on_SmtpUTF8Mailbox = 1208;
```

Comment:
can you include comments where all these magic numbers came from?

Reply:
fix on https://github.com/oven-sh/bun/commit/927f54269bdf489d21be4a8a6b89b448fdf48184